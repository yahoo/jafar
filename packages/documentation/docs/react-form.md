---
id: react-form
title: Form
sidebar_label: Form
---

Form component gets as props form model object and resources object that will list all its settings (such as validations, disable term, field dependencies and more) and lifecycle hooks (such as before and after actions, submit, form level validations and dto conversions). In addition, Form component exposes a context object that includes current form model, resources and actions (such as `changeValue`) to perform on the form. The context can be used within any child component of the Form component (not necessary direct child component). 

## Props

| Name          | Type          | Description |
| ------------- |-------------| ------------|
| model      | required object | Model object that defines fields, validators and more as described in [Form Class](arguments#model) |
| resources      | object      | Resources object that defines all the handlers that the form model defined as described in [Form Class](arguments#resources) |
| settings      | object | Settings overrides object as described in [Form Class](arguments#settings) |
| data      | object      | Pass [data](data) as a prop outside the model object - only when data might change outside the Form component during the app life. On `init` - Form will applies data to the model and pass the final model to the underline [Form class](form-overview) instance, and when data prop reference changes only `changeData` action called on the underline form instance. |
| context      | object      | Pass [context](context) as a prop outside the model object - only when context might change outside the Form component during the app life. On `init` - Form will applies context to the model and pass the final model to the underline [Form class](form-overview) instance, and when context prop reference changes only `changeContext` action called on the underline form instance. |

## Example

```jsx
import { Form, Field } from '@jafar/react-form';
import SomeInputText from './my-components/Text.jsx';

const model = {
  id: 'user-form',
  fields: {
    firstName: {
      path: 'firstName',
      label: 'First Name',
      description: 'Enter user first name',
      component: {
        name: 'InputText',
      },
    },
    lastName: {
      path: 'lastName',
      label: 'Last Name',
      component: {
        name: 'InputText',
      },
    },
    data: {
      firstName: 'Roos',
      lastName: 'Geller',
    }
  },
};

const resources = {
  components: {
    InputText: { renderer: SomeInputText },
  }
};

<Form model={model} resources={resources}>
  <h2>User Details</h2>
  <Field id="firstName" />
  <Field id="lastName" />
</Form>
```

## Context

Form provides a context object used by direct and non-direct children components. 

> **Note:** In `@jafar/react-form` - `context` refer to the data provided by the `Form` component object, which can be consumed by each
underline component (i.e react's context feature). Whereas in `@jafar/form` [context](context) refers to the app data which is stored in the model to
serve the form's lifecycle handlers.

### Context object 

Context object contains: 

| Name          | Type          | Description |
| ------------- |-------------| ------------|
| model         | object | The current model object. Can be used to pull data for an underline component. For instance, a save button component, should use form data to disable the button in some cases |
| resources     | object      | The current resources object |
| actions       | object      | actions object contains a set of function to be called in order to manipulate the from. Full actions docs can be found [here](actions.html) |
| parent       | object      | parent form [context](react-form#context-object) object |

For Example:

```javascript
const reset = () => {
  context.actions.reset();
};

<Button disabled={context.model.invalid || context.model.processing || 
  !context.model.dirty} onClick={reset}>Reset</Button>
```

### Context Consume

In order to use the Form context in child components, components first need to consume Form context.

Example using class component:

```javascript
import React from 'react';
import { FormContext } from '@jafar/react-form';

class ErrorsSummary extends React.Component {
  static contextType = FormContext;

  render() {
    return (<div>
      <h4>Errors summary:</h4>
      {
          Object.values(this.context.model.fields).map((field, i) => (<ul key={i}>
            { field.errors.map((error, index) => <li key={index}>Error in {field.label}: {error.message}</li>) }
          </ul>))
      }
    </div>);
  }
}
```

Example using functional component:

```javascript
import React, { useContext } from 'react';
import { FormContext } from '@jafar/react-form';

const ErrorsSummary = () => {
  const context = useContext(FormContext);

  return (<div>
    <h4>Errors summary:</h4>
    {
        Object.values(context.model.fields).map((field, i) => (<ul key={i}>
          { field.errors.map((error, index) => <li key={index}>Error in {field.label}: {error.message}</li>) }
        </ul>))
    }
  </div>);
};
```

### Context Usage

Components declared inside a Form tag component can define contextType of form, and then
use `this.context` to gain access to the form context.

Examples
1. Jafar [Field component](componentsField.html) - consumes the form context and interacts with it. It uses the
`this.context.model` to get field's data, `this.context.actions.changeValue` and `this.context.actions.changeState` actions to notify the form about a change in the component.
2. Your custom component. You can define another custom components to interact with the form like `SaveButton`

Example

```javascript
import React from 'react';
import Button from '@material-ui/core/Button';
import { FormContext } from '@jafar/react-form';

export default class SaveButton extends React.Component {
  static contextType = FormContext;

  render() {
    return (<Button disabled={this.context.model.invalid} onClick={this.onSave}></Button>);
  }

  onSave = () => {
    // save data to the server
    serverCallToSaveData(this.context.model.data).then(() => {
      // reset the form to its initial state
      this.context.actions.reset();
    });
  }
}
```

3. Pass to custom components any additional data (for example app data) through resources or model objects that are passed via context.

Example

```javascript
const model = {
  // ...
  appData: {
    user: { /* ... */ },
    currency: 'USD',
  },
}

export default class Currency extends React.Component {
  static contextType = FormContext;

  render() {
    return (<Input
      type="number" 
      value={this.props.value} 
      onChange={this.onValueChange} 
      endAdornment={<Label value={this.context.model.appData.currency}/>}/>);
  }

  onValueChange = (e) => {
    this.props.onValueChange(e.target.value);
  }
}
```

## Using createForm helper
Instead of defining multiple custom inner components using form context (like the above SaveButton), you can use our `createForm` function. This function
gets a form definition (model & resources) and returns an HOC that gets your single component. Your component should define its context as the form context, and then the form context will be available in your component.

Example
```javascript
import { Field, FormContext, createForm } from '@jafar/react-form';
import Button from '@material-ui/core/Button';
import form from '../form'; // object of { model, resources }

class MyComponent extends React.Component {
  static contextType = FormContext;

  render() {
    return (<div>
      <h2>User Details</h2>
      <Field id="firstName" />
      <Field id="lastName" />
      <Button disabled={this.context.model.invalid} onClick={this.onSave}></Button>
    </div>
     );
  }

  onSave = () => {
    // save data to the server
    serverCallToSaveData(this.context.model.data).then(() => {
      // reset the form to its initial state
      this.context.actions.reset();
    });
  }
}

return createForm(form)(MyComponent);
```
