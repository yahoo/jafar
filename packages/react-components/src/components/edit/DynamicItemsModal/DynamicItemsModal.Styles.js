/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import styled from 'styled-components';

const ItemsListContainer = styled.ul`
	list-style: none;
	padding: 0;
	max-height: 180px;
	overflow-y: auto;
`;

const ListItem = styled.li`
	display: flex;
	min-height: 40px;
	position: relative;
	&:hover div:nth-child(2){
		visibility: visible;
	}
`;

const ItemRendererContainer = styled.div`
	justify-content: flex-start;
	flex: 1;
	white-space: nowrap;
	overflow: hidden;
	text-overflow: ellipsis;
`;

const ActionsContainer = styled.div`
	justify-content: flex-end;
	visibility: hidden;
	position: absolute;
	right: 0;
	top: 0;
`;

const AddLabel = styled.span`
	text-transform: uppercase;
	margin-left: 10px;
	font-size: 16px;
	line-hight: 16px;
	cursor: default;
`;

export default {
  ItemsListContainer,
  ItemRendererContainer,
  ActionsContainer,
  ListItem,
  AddLabel,
};
