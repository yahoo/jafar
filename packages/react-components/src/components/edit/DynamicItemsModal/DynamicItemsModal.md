<h3>Examples</h3>

<h4>Simple</h4>

```javascript
const useContext = require('react').useState;
const createContext = require('react').createContext;
const JafarContext = require('@jafar/react-form/context').default;
const InputText = require('../Text').default;
const Form = require('@jafar/react-form/Form').default;
const Field = require('@jafar/react-form/Field').default;

// jafar form definition
const employeeForm = {
  model: {
    id: 'employees',
      fields: { 
        firstName: {
          label: 'First name',
          path: 'firstName',
          component: {
              name: 'InputText',
          },
          required: true,
        },
        lastName: {
          label: 'Last name',
          path: 'lastName',
          component: {
              name: 'InputText',
          },
          required: true,
        },
      },
  },
  resources: {
    components: { 
      InputText: {
        renderer: InputText,
      } 
    },
  }
};

// jafar-layout item
const employeeFormLayout = { 
  size: 1,
  sections: [{ 
      id: 'AddEmployee', 
      boxes: [{
          direction: 'column',
          style: { 'margin': '20px' },
          boxes: [{
              component: Field,
              props: { id: 'firstName' },
          }, {
              component: Field,
              props: { id: 'lastName' },
          }],
      }],
  }],
}

// render single item
const employeeItemRendered = (item) => (<span>{item.firstName} {item.lastName}</span>);

// main form
const model = {
  id: 'company',
  fields: {
    employees: {
      label: 'Employees',
      path: 'employees',
      component: {
        name: 'DynamicItemsModal',
        state: {
          resourceId: 'employeesData',
        },
      },
      required: true,
    },
  },
  data: {
    employees: [{
      firstName: 'Rachel',
      lastName: 'Green',
    }, {
      firstName: 'Ross',
      lastName: 'Geller',
    }]
  },
};

const resources = {
  components: { 
    DynamicItemsModal: {
      renderer: DynamicItemsModal,
    },
  },
  employeesData: { // custom data
      form: employeeForm,
      item: employeeFormLayout,
      itemRenderer: employeeItemRendered,
  },
};

<div style={{ width: '400px' }}>
  <Form model={model} resources={resources}>
    <Field id='employees' />
  </Form>
</div>
  ```
<h4>Advanced</h4>

```javascript
const useContext = require('react').useState;
const createContext = require('react').createContext;
const JafarContext = require('@jafar/react-form/context').default;
const InputText = require('../Text').default;
const Select = require('../Select').default;
const Form = require('@jafar/react-form/Form').default;
const Field = require('@jafar/react-form/Field').default;

const selectItems = {
  M: { value: 'M', label: 'Male'},
  F: { value: 'F', label: 'Female'},
};
// jafar form 
const employeeForm = {
  model: {
    id: 'employees',
    fields: { 
      firstName: {
        label: 'First name',
        path: 'firstName',
        component: {
            name: 'InputText',
        },
        required: true,
      },
      lastName: {
        label: 'Last name',
        path: 'lastName',
        component: {
            name: 'InputText',
        },
        required: true,
      },
      gender: {
        label: 'Gender',
        path: 'gender',
        component: {
          name: 'Select',
          state: {
            items: Object.values(selectItems),
          },
        },
      },
      address: {
        label: 'Address',
        path: 'address',
        component: {
            name: 'InputText',
        },
        required: true,
      },
    },
    data: { gender: 'F' },
  },
  resources: {
    components: { 
      InputText: {
        renderer: InputText,
      }, 
      Select: {
        renderer: Select,
      }
    },
  },
};

// jafar-layout item
const employeeFormLayout = {
  size: 1,
  sections: [{ 
      id: 'basic', 
      title: 'Basic',
      boxes: [{
          direction: 'column',
          boxes: [{
              component: Field,
              props: { id: 'firstName' },
          }, {
              component: Field,
              props: { id: 'lastName' },
          }, {
              component: Field,
              props: { id: 'gender' },
          }],
      }],
  }, { 
      id: 'more', 
      title: 'More',
      boxes: [{
          direction: 'column',
          boxes: [{
              component: Field,
              props: { id: 'address' },
          }],
      }],
  }],
};

const employeeItemRenderer = item => (<div>
  <div>{item.firstName} {item.lastName}</div>
  <div style={{ paddingLeft: '20px' }}>
    <div>üë§ {selectItems[item.gender].label}</div>
    <div>üè† {item.address}</div>
  </div> 
</div>);

// main form
const model = {
  id: 'company',
  fields: {
    employees: {
      label: 'Employees',
      path: 'employees',
      component: {
        name: 'DynamicItemsModal',
        state: {
            resourceId: 'employeesItems',
            addButtonLabel: 'Add New',
            addModalTitle: 'Add employee',
            editModalTitle: 'Edit employee',
            saveButtonLabel: 'Save',
            cancelButtonLabel: 'Cancel',
        },
      },
    },
  },
  data: {
    employees: [{
      firstName: 'Rachel',
      lastName: 'Green',
      gender: 'F',
      address: '90 Bedford St, New York, NY 10014, USA',
    }, {
      firstName: 'Ross',
      lastName: 'Geller',
      gender: 'M',
      address: '90 Bedford Street',
    }]
  },
};

const resources = {
  components: { 
    DynamicItemsModal: {
      renderer: DynamicItemsModal,
    },
  },
  employeesItems: {
    form: employeeForm,
    item: employeeFormLayout,
    itemRenderer: employeeItemRenderer,
  },
};

<div style={{ width: '600px' }}>
  <Form model={model} resources={resources}>
    <Field id='employees' />
  </Form>
</div>
  ```
