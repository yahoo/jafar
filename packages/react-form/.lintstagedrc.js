module.exports = require('@jafar/react-kit/config/lintstagedrc.js');