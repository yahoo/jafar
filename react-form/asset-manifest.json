{
  "files": {
    "main.css": "https://yahoo.github.io/jafar/react-form/static/css/main.99903ae4.css",
    "main.js": "https://yahoo.github.io/jafar/react-form/static/js/main.cf9d777f.js",
    "main.css.map": "https://yahoo.github.io/jafar/react-form/static/css/main.99903ae4.css.map",
    "main.js.map": "https://yahoo.github.io/jafar/react-form/static/js/main.cf9d777f.js.map",
    "index.html": "https://yahoo.github.io/jafar/react-form/index.html",
    "precache-manifest.0aa3954ea0dd693b697b5c24c7bb696a.js": "https://yahoo.github.io/jafar/react-form/precache-manifest.0aa3954ea0dd693b697b5c24c7bb696a.js",
    "service-worker.js": "https://yahoo.github.io/jafar/react-form/service-worker.js",
    "static/js/main.cf9d777f.js.LICENSE": "https://yahoo.github.io/jafar/react-form/static/js/main.cf9d777f.js.LICENSE",
    "static/media/jafar.svg": "https://yahoo.github.io/jafar/react-form/static/media/jafar.1c1a5205.svg"
  },
  "entrypoints": [
    "static/css/main.99903ae4.css",
    "static/js/main.cf9d777f.js"
  ]
}