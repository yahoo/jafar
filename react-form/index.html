<!doctype html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width,initial-scale=1,shrink-to-fit=no"><meta name="theme-color" content="#000000"><link rel="manifest" href="https://yahoo.github.io/jafar/react-form/manifest.json"><link rel="shortcut icon" href="https://yahoo.github.io/jafar/react-form/favicon.ico"><title>Jafar | React Form Demos</title><link href="https://yahoo.github.io/jafar/react-form/static/css/main.99903ae4.css" rel="stylesheet"></head><body><noscript>You need to enable JavaScript to run this app.</noscript><div id="root"></div><script src="https://yahoo.github.io/jafar/react-form/static/js/main.8c061ea5.js"></script></body></html>