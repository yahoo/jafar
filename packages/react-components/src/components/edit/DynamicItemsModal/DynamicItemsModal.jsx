/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import Item from '@jafar/react-layout/Item';
import { clone } from 'lodash';
import AddIcon from '@material-ui/icons/Add';
import RemoveIcon from '@material-ui/icons/Delete';
import EditIcon from '@material-ui/icons/Create';
import Button from '@material-ui/core/Button';
import Fab from '@material-ui/core/Fab';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import DialogTitle from '@material-ui/core/DialogTitle';
import { Form, FormContext } from '@jafar/react-form';
import Styled from './DynamicItemsModal.Styles';

const deafultLabels = {
  addModalTitle: 'Add',
  editModalTitle: 'Edit',
  addButtonLabel: 'Add',
  saveButtonLabel: 'Save',
  cancelButtonLabel: 'Cancel',
};

/**
 * Represent array of objects. Can add / edit / remove items from the array
 * 
 * Import <a target="_blank" href=
 "https://github.com/yahoo/jafar/blob/master/packages/react-components/src/components/edit/
 DynamicItemsModal/DynamicItemsModal.jsx">
 DynamicItemsModal</a> from '@jafar/react-components/edit/DynamicItemsModal'
 */
class DynamicItemsModal extends React.Component {
  static contextType = FormContext;

  static propTypes = {
    value: PropTypes.array,
    state: PropTypes.shape({
      resourceId: PropTypes.string.isRequired,
      addModalTitle: PropTypes.string,
      editModalTitle: PropTypes.string,
      addButtonLabel: PropTypes.string,
      saveButtonLabel: PropTypes.string,
      cancelButtonLabel: PropTypes.string,
    }).isRequired,
    onValueChange: PropTypes.func.isRequired,
    onStateChange: PropTypes.func,
  };

  static defaultProps = {
    value: [],
    state: {
    },
  };
 
  constructor(props, context) {
    super(props);
    this.resource = getResource(this.props.state, context);
    this.initialData = this.resource.form.model.data || {};
  }

  open = () => {
    const state = clone(this.props.state);
    this.dialogTitle = state.addModalTitle || deafultLabels.addModalTitle;
    state.isOpen = true;
    this.props.onStateChange(state);
  };

  edit = index => {
    const state = clone(this.props.state);
    const item = this.props.value[index];
    this.resource.form.model.data = { ...item };
    this.dialogTitle = state.editModalTitle || deafultLabels.editModalTitle;
    state.isOpen = true;
    state.isEditMode = true;
    state.editIndex = index;
    this.props.onStateChange(state);
  };
    
  add = item => {
    const value = clone(this.props.value);
    const state = clone(this.props.state);
    if (state.isEditMode) {
      value[state.editIndex] = item;
      this.resource.form.model.data = this.initialData;
    } else {
      value.push(item);
    }
    this.props.onValueChange(value);
    this.close();
  };
    
  remove = index => {
    const value = clone(this.props.value);
    value.splice(index, 1);
    this.props.onValueChange(value);
  };
    
  close = () => {
    const state = clone(this.props.state);
    this.resource.form.model.data = this.initialData;
    state.isOpen = false;
    state.isEditMode = false;
    delete state.editIndex;
    this.props.onStateChange(state);
  };

  renderItemsDisplay = () => {
    return (<Styled.ItemsListContainer>
      {
        this.props.value.map((item, index) => 
          (<Styled.ListItem key={index}>
            <Styled.ItemRendererContainer>{this.resource.itemRenderer(item, index)} </Styled.ItemRendererContainer>
            <Styled.ActionsContainer>
              <Fab style={{ margin: '0 5px' }} size="small" color="secondary" onClick={() => { this.edit(index); }}>
                <EditIcon />
              </Fab>
              <Fab style={{ margin: '0 5px' }} size="small" color="secondary" onClick={() => { this.remove(index); }}>
                <RemoveIcon />
              </Fab>
            </Styled.ActionsContainer>
          </Styled.ListItem>))
      }
    </Styled.ItemsListContainer>);
  };

  render() {
    return (<div>
      {this.renderItemsDisplay()}
      <Fab size="small" color="primary" onClick={this.open}>
        <AddIcon />
      </Fab>
      <Styled.AddLabel>{this.props.state.addButtonLabel || deafultLabels.addButtonLabel}</Styled.AddLabel>

      {this.props.state.isOpen && (
        <Form model={this.resource.form.model} resources={this.resource.form.resources}>
          <DialogForm 
            item={this.resource.item} 
            add={this.add} 
            close={this.close}
            title={this.dialogTitle}
            isOpen={this.props.state.isOpen}
            saveButtonLabel={this.props.state.saveButtonLabel || deafultLabels.saveButtonLabel}
            cancelButtonLabel={this.props.state.cancelButtonLabel || deafultLabels.cancelButtonLabel} />
        </Form>)}
    </div>);
  }
}

const DialogForm = ({ item, isOpen, add, close, title, saveButtonLabel, cancelButtonLabel }) => {
  const formContext = useContext(FormContext);

  return (<Dialog open={!!isOpen} onClose={close}>
    <DialogTitle>{title}</DialogTitle>
    <DialogContent>
      <Item {...item} />
    </DialogContent>
    <DialogActions>
      <Button onClick={close} color="primary">{cancelButtonLabel}</Button>
      <Button onClick={() => add(formContext.model.data)} color="primary" 
        disabled={formContext.model.invalid} autoFocus={true}>{saveButtonLabel}</Button>
    </DialogActions>
  </Dialog>);
};


function getResource (state, context) {
  const key = state && state.resourceId ? state.resourceId : '';
  return context.resources[key];
}

export default DynamicItemsModal;
