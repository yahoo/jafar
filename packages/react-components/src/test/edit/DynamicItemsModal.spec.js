/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import { mount, shallow, render } from 'enzyme';
import { FormContext, Field } from '@jafar/react-form';
import InputText from '../../components/edit/Text';
import DynamicItemsModal from '../../components/edit/DynamicItemsModal/index';

describe('<DynamicItemsModal />', () => {
  let component;
  let onValueChangeSpy;
  let onStateChangeSpy;
  let value;
  let state;
  let model;
  let resources;
  beforeEach(() => {
    onValueChangeSpy = jest.fn();
    onStateChangeSpy = jest.fn();
    value = [];
    state = {
      resourceId: 'employeesItems',
      addButtonLabel: 'Add New',
      addModalTitle: 'Add employee',
      editModalTitle: 'Edit employee',
    };
    model = {
      id: 'company',
      fields: {
        employees: {
          label: 'Employees',
          path: 'employees',
          component: {
            name: 'DynamicItemsModal',
            state: {
              resourceId: 'employeesItems',
              addButtonLabel: 'Add New',
              addModalTitle: 'Add employee',
              editModalTitle: 'Edit employee',
            },
          },
        },
      },
    };
    resources = {
      components: { DynamicItemsModal },
      employeesItems: {
        form: { // form
          model: {
            id: 'employees',
            fields: { 
              firstName: {
                label: 'First name',
                path: 'firstName',
                component: {
                  name: 'InputText',
                },
                required: true,
              },
              lastName: {
                label: 'Last name',
                path: 'lastName',
                component: {
                  name: 'InputText',
                },
                required: true,
              },
            },
          },
          resources: {
            components: { InputText },
          },
        },
        item: { // react-layout item
          title: 'Simple',
          sections: [{ 
            id: 'Add employee', 
            boxes: [{
              direction: 'column',
              boxes: [{
                component: Field,
                props: { id: 'firstName' },
              }, {
                component: Field,
                props: { id: 'lastName' },
              }],
            }],
          }],
        },
        itemRenderer: (item) => (<span>{item.firstName} - {item.lastName}</span>),
      },
    };
  });

  it('should render provided data', () => {
    component = render(getComponent(value, state, onStateChangeSpy, onValueChangeSpy, model, resources));
    expect(component).toMatchSnapshot();
  });
  describe('component functions', () => {
    it('call open method', () => {
      component = mount(getComponent(value, state, onStateChangeSpy, onValueChangeSpy, model, resources));
      const instance = component.instance();
      instance.open();
      expect(onStateChangeSpy).toHaveBeenCalledWith({
        addButtonLabel: 'Add New',
        addModalTitle: 'Add employee',
        editModalTitle: 'Edit employee', 
        isOpen: true,
        resourceId: 'employeesItems',
      });
      expect(instance.dialogTitle).toBe(state.addModalTitle);
    });
    it('call close method', () => {
      component = mount(getComponent(value, state, onStateChangeSpy, onValueChangeSpy, model, resources));
      const instance = component.instance();
      instance.close();
      expect(onStateChangeSpy).toHaveBeenCalledWith({
        addButtonLabel: 'Add New',
        addModalTitle: 'Add employee',
        editModalTitle: 'Edit employee',  
        isEditMode: false,
        isOpen: false,
        resourceId: 'employeesItems',
      });
    });
    it('call add method, when adding new item', () => {
      const item = { firstName: 'first', lastName: 'last' };
      component = mount(getComponent(value, state, onStateChangeSpy, onValueChangeSpy, model, resources));
      const instance = component.instance();
      instance.add(item);
      expect(onValueChangeSpy).toHaveBeenCalledWith([item]);
      expect(onStateChangeSpy).toHaveBeenCalledWith({
        addButtonLabel: 'Add New',
        addModalTitle: 'Add employee',
        editModalTitle: 'Edit employee', 
        isEditMode: false,
        isOpen: false,
        resourceId: 'employeesItems',
      });
    });
    it('call add method, when editing new item', () => {
      const item = { firstName: 'edit first', lastName: 'edit last' };
      state.isEditMode = true; 
      state.editIndex = 0;
      value = [{ firstName: 'first', lastName: 'last' }];
      component = mount(getComponent(value, state, onStateChangeSpy, onValueChangeSpy, model, resources));
      const instance = component.instance();
      instance.add(item);
      expect(onValueChangeSpy).toHaveBeenCalledWith([item]);
      expect(onStateChangeSpy).toHaveBeenCalledWith({
        addButtonLabel: 'Add New',
        addModalTitle: 'Add employee',
        editModalTitle: 'Edit employee',
        isEditMode: false,
        isOpen: false,
        resourceId: 'employeesItems',
      });
    });
    it('call edit method', () => {
      value = [{ firstName: 'first', lastName: 'last' }];
      component = mount(getComponent(value, state, onStateChangeSpy, onValueChangeSpy, model, resources));
      const instance = component.instance();
      instance.edit(0);
      expect(instance.context.resources.employeesItems.form.model.data).toEqual(value[0]);
      expect(onStateChangeSpy).toHaveBeenCalledWith({
        addButtonLabel: 'Add New',
        addModalTitle: 'Add employee',
        editModalTitle: 'Edit employee',
        editIndex: 0,
        isEditMode: true,
        isOpen: true,
        resourceId: 'employeesItems',
      });
      expect(instance.dialogTitle).toBe(state.editModalTitle);
    });
    it('call remove method', () => {
      value = [{ firstName: 'first', lastName: 'last' }];
      component = mount(getComponent(value, state, onStateChangeSpy, onValueChangeSpy, model, resources));
      const instance = component.instance();
      instance.remove(0);
      expect(onValueChangeSpy).toHaveBeenCalledWith([]);
    });
    it('render itemsDisplay', () => {
      value = [{ firstName: 'first', lastName: 'last' }];
      component = mount(getComponent(value, state, onStateChangeSpy, onValueChangeSpy, model, resources));
      const instance = component.instance();
      const itemsDisplay = instance.renderItemsDisplay(); 
      expect(shallow(itemsDisplay)).toMatchSnapshot();
    });
  });
  function getComponent(value, state, onStateChange, onValueChange, model, resources) {
    return (<FormContext.Provider value={{ model, resources }}>
      <DynamicItemsModal 
        value={value}
        state={state}
        onStateChange={onStateChange}
        onValueChange={onValueChange} />
    </FormContext.Provider>);
  }
});
