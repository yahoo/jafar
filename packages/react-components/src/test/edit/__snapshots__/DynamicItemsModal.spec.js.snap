// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DynamicItemsModal /> component functions render itemsDisplay 1`] = `
<StyledComponent
  forwardedComponent={
    Object {
      "$$typeof": Symbol(react.forward_ref),
      "attrs": Array [],
      "componentStyle": ComponentStyle {
        "componentId": "sc-gqjmRU",
        "isStatic": false,
        "lastClassName": "jbSsaT",
        "rules": Array [
          "
	list-style: none;
	padding: 0;
	max-height: 180px;
	overflow-y: auto;
",
        ],
      },
      "displayName": "styled.ul",
      "foldedComponentIds": Array [],
      "render": [Function],
      "styledComponentId": "sc-gqjmRU",
      "target": "ul",
      "toString": [Function],
      "warnTooManyClasses": [Function],
      "withComponent": [Function],
    }
  }
  forwardedRef={null}
>
  <styled.li
    key="0"
  >
    <styled.div>
      <span>
        first
         - 
        last
      </span>
       
    </styled.div>
    <styled.div>
      <WithStyles(ForwardRef(Fab))
        color="secondary"
        onClick={[Function]}
        size="small"
        style={
          Object {
            "margin": "0 5px",
          }
        }
      >
        <CreateIcon />
      </WithStyles(ForwardRef(Fab))>
      <WithStyles(ForwardRef(Fab))
        color="secondary"
        onClick={[Function]}
        size="small"
        style={
          Object {
            "margin": "0 5px",
          }
        }
      >
        <DeleteIcon />
      </WithStyles(ForwardRef(Fab))>
    </styled.div>
  </styled.li>
</StyledComponent>
`;

exports[`<DynamicItemsModal /> should render provided data 1`] = `
<div>
  <ul
    class="sc-gqjmRU jbSsaT"
  />
  <button
    class="MuiButtonBase-root MuiFab-root MuiFab-sizeSmall MuiFab-primary"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiFab-label"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
        />
      </svg>
    </span>
  </button>
  <span
    class="sc-jzJRlG hfSEMm"
  >
    Add New
  </span>
</div>
`;
